In [1]:
import random
import math
import numpy as np

In [2]:
dice = [1, 2, 3, 4, 5, 6]

# Testing randomness

Function `random.choice(data: list)` returns random value from `data`.

<br>

Computer use an initial value called **seed** to start the number generation process. Usually algorithms uses **system's current time**, but exists thousands of methods to generate random numbers.

In [3]:
for num_test in range(10):
    result = ''

    for attempt in range(10):
        result = result + str(random.choice(dice))

    print(result)

6126415645
5126552522
1562213516
3236623222
5161664132
2354126543
4136515662
3525653325
1223424651
5341325252


In [4]:
for num_test in range(10):
    random.seed(0)

    result = ''

    for attempt in range(10):
        result = result + str(random.choice(dice))

    print(result)

4413544343
4413544343
4413544343
4413544343
4413544343
4413544343
4413544343
4413544343
4413544343
4413544343


# A simulation of Die Rolling

We want to simulate approximated probability that rolling five times we got combination `11111`.

<br>

We can calculate it using **multiplicative law**.

<br>

$\mathcal{P}(x=11111)=\frac{1}{6} \times \frac{1}{6} \times \frac{1}{6} \times \frac{1}{6} \times \frac{1}{6}=\frac{1}{7776} \approx$ `0.0001286`

<br>

We can prove it using **brute force algorithm**.

In [5]:
guess = '11111' # You can modify this value to check results for another combinations

Assume that $n=$ size of a dice, and $k=$ size of guess.

### Hand-written algorithm

In [6]:
num_guessed = 0
num_tests = 0

result = ''

for first_digit in dice:
    for second_digit in dice:
        for third_digit in dice:
            for fourth_digit in dice:
                for fifth_digit in dice:
                    result = result + str(first_digit) + str(second_digit) + str(third_digit) + str(fourth_digit) + str(fifth_digit)
                    if result == guess:
                        num_guessed = num_guessed + 1
                    num_tests = num_tests + 1
                    result = ''

print('Guessed:', num_guessed)
print('Tests:', num_tests)
print('Probability:', num_guessed / num_tests)

Guessed: 1
Tests: 7776
Probability: 0.0001286008230452675


Time Complexity: $\mathcal{O}(n^5)$, and \
Space Complexity: $\mathcal{O}(n)$.

### Recursive algorithm

In [7]:
num_guessed = 0
num_tests = 0

result = ''

def recursive_algorithm(dice, guess):
    recursive_loop(dice, guess, '', 0)


def recursive_loop(dice, guess, current_result, depth):
    global num_guessed, num_tests
    if depth == len(guess):
        if current_result == guess:
            num_guessed = num_guessed + 1
        num_tests = num_tests + 1
    else:
        for digit in dice:
            recursive_loop(dice, guess, current_result + str(digit), depth + 1)


recursive_algorithm(dice, guess)

print('Guessed:', num_guessed)
print('Tests:', num_tests)
print('Probability:', num_guessed / num_tests)

Guessed: 1
Tests: 7776
Probability: 0.0001286008230452675


Time Complexity: $\mathcal{O}(n^k)$, and \
Space Complexity: $\mathcal{O}(n^k)$.

In cases where the number of values is small, the brute force algorithm is not a bad idea. However, when analyzing much more complex cases, it may prove to be too time-consuming for a computer. Therefore, in probability studies, algorithms based on **randomness** are often used.

### Stochastic Algorithm

In [8]:
num_trials = 1000
num_guessed = 0

for trial in range(num_trials):
    result = ''

    for attempt in range(len(guess)):
        result = result + str(random.choice(dice))
    
    if result == guess:
        num_guessed = num_guessed + 1


print('Guessed:', num_guessed)
print('Trials:', num_trials)
print('Probability:', num_guessed / num_trials)

Guessed: 0
Trials: 1000
Probability: 0.0


Time Complexity: $\mathcal{O}(n^2)$ or maybe $\mathcal{O}(n^3)$, because of `random.choice(data: list)`, and \
Space Complexity: $\mathcal{O}(n)$.

It should be noted that simulation does not provide a 100% accurate result. However, it can be very useful when calculating complex processes. Also, the bigger number of trials, the better approximation we get.

In [9]:
num_trials = 1000000
num_guessed = 0

for trial in range(num_trials):
    result = ''

    for attempt in range(len(guess)):
        result = result + str(random.choice(dice))
    
    if result == guess:
        num_guessed = num_guessed + 1


print('Guessed:', num_guessed)
print('Trials:', num_trials)
print('Probability:', num_guessed / num_trials)

Guessed: 128
Trials: 1000000
Probability: 0.000128


| Algorithm                               | Probability          |
|-----------------------------------------|----------------------|
| Hand-written algorithm                  | 0.0001286008230452675|
| Recursive algorithm                     | 0.0001286008230452675|
| Stochastic algorithm for 1000 trials    | 0.0                  |
| Stochastic algorithm for 1000000 trials | 0.000128             |


It can be observed that the runtime of this algorithm was the longest. However, this should not be a cause for concern, as the example is simple enough that it does not benefit much from a randomness-based approach. This does not mean, however, that the algorithm is useless for such simple computations. On the contrary, in the case of highly complex probabilities involving numerous scenarios, these algorithms become invaluable, even if they require greater computational power and potentially yield slightly less precise results.

# The power of simulation on example of the birthday problem

**What is the probability that in a group of $n < 365$ at leats $k < n$ people share the same day and month of birth?**

<br>

The better way to calculate is to solve the opposite problem, when no one shares a birthday day and month using the following formula. 
$$\mathcal{P}(X) = 1 - \neg\mathcal{P}(X)$$
The birthday problem for 2 people would looks like

$\mathcal{P}($ at least two people share a birthday $) = 1 - \mathcal{P}( $ no one shares a birthday $ ) $.

The common knowledge of this problem is that $\mathcal{P} ( $ no one shares a birthday $ ) = \frac{365!}{(365 - n)! \times 365^n}$.

<br>

The conclusion is the following formula.

$\mathcal{P} ( $ at leats two people share a birthday $ ) = 1 - \frac{365!}{(365 - n)! \times 365^n}$

<br>

Let's compare the simulation with the real result.

In [10]:
n = 10 # You can modify this value to check results for another combinations

In [11]:
real_probability = 1 - ((math.factorial(365)) / (math.factorial(365 - n) * (365 ** n)))

num_trials = 10000
num_guessed = 0

birthday_dates = [ day for day in range(365) ]

for trial in range(num_trials):
    calendar = [ 0 for day in range(365) ]

    for group in range(n):
        random_day = random.choice(birthday_dates)
        calendar[random_day] = calendar[random_day] + 1
    
    for birthdays in calendar:
        if birthdays >= 2:
            num_guessed = num_guessed + 1
            break

print('Approximated Probability:', num_guessed / num_trials)
print('Real Probability:', real_probability)

Approximated Probability: 0.1147
Real Probability: 0.11694817771107768


This seems very simple to compute mathematically. However, the situation becomes much more challenging when $k = 10$. In such a case, you need to calculate \
$\mathcal{P} ($ no one shares a birthday $ ) $, $\mathcal{P} ($ exactly one pair shares a birthday $ ) $, $\mathcal{P} ($ exactly two pairs shares a birthday $ ) $,  and so on.

This is why simulations are so helpful for complex processes. For instance, when probabilities related to biological, chemical, or physical processes are involved, formulating a probability formula may prove to be too complicated.

Below you will find a ready-made program where you can set the group size, the number of people sharing the same birthday, and the number of experiments to be conducted. Remember, the more experiments you run, the more accurate the result will be.

In [12]:
k = 2 # You can modify this value to check results for another combinations
n = 10 # You can modify this value to check results for another combinations
num_trials = 100000 # You can modify this value to check results for another combinations

In [13]:
num_guessed = 0

birthday_dates = [ day for day in range(365) ]

for trial in range(num_trials):
    calendar = [ 0 for day in range(365) ]

    for group in range(n):
        random_day = random.choice(birthday_dates)
        calendar[random_day] = calendar[random_day] + 1
    
    for birthdays in calendar:
        if birthdays >= k:
            num_guessed = num_guessed + 1
            break

print('Approximated Probability:', num_guessed / num_trials)

Approximated Probability: 0.11521
